### PR TITLE
feat: add server module to abstract grpc, http and a muxed server

### PR DESCRIPTION
**Pull request by rohilsurana**
Supports -
✅ GRPC server
✅ HTTP server
✅ Muxed Server
✅ GRPC Middleware
✅ Graceful shutdown
✅ GRPC Health

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,19 +5,24 @@ go 1.16
 require (
 	github.com/charmbracelet/glamour v0.3.0
 	github.com/gorilla/mux v1.8.0
-	github.com/grpc-ecosystem/grpc-gateway/v2 v2.6.0 // indirect
+	github.com/grpc-ecosystem/go-grpc-middleware v1.3.0
+	github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
```
Comment (whoAbhishekSah) [inline go.mod:9]: Did we remove new-relic in favor of Prometheus?

File: server/example/main.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	grpc_prometheus.UnaryServerInterceptor,
+	grpc_zap.UnaryServerInterceptor(zap.NewExample()),
+))
+
+func main() {
+	grpcPort := 8000
+	httpPort := 8080
+	muxPort := 9000
+	gatewayClientPort := grpcPort
+	go grpcS(grpcPort)
```
Comment (whoAbhishekSah) [inline server/example/main.go:37]: Let's use the fully qualified name? Like `grpcS` to be called `grpcServer`
Similarly for `httpS` and `muxS`

File: server/mux.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	}
+}
+
+func WithMuxHTTPServer(httpServer *http.Server) MuxOption {
+	return func(mos *muxOptions) {
+		WithHTTPServer(httpServer)(&mos.httpOptions)
+	}
+}
+
+func NewMux(config Config, options ...MuxOption) (*MuxServer, error) {
```
Comment (krtkvrm) [inline server/mux.go:45]: why is cmux being added? we can just use different port for grpc & http

File: server/grpc.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+func (s *GRPCServer) Shutdown(ctx context.Context) {
+	s.grpcServer.GracefulStop()
+	go func() {
+		<-ctx.Done()
+		s.grpcServer.Stop()
+	}()
+}
+
+func (s *GRPCServer) RegisterHealth() *health.Server {
+	hs := health.NewServer()
```
Comment (kushsharma) [inline server/grpc.go:76]: I have not used it so not sure but don't we need to implement this health server?

File: server/example/main.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	s.Shutdown(shutdownCtx)
+}
+
+func muxS(muxPort int) {
+	ctx, cancelFunc := context.WithCancel(server.HandleSignals(context.Background()))
+	defer cancelFunc()
+
+	s, err := server.NewMux(server.Config{
+		Port: muxPort,
+		Host: "",
```
Comment (krtkvrm) [inline server/example/main.go:103]: no need for `Host: ""`, as if not assigned, it will be empty string by default. similar for other occurrences of this

**Review by kushsharma — APPROVED** [review]

File: server/sys.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+func HandleSignals(ctx context.Context) context.Context {
+	newCtx, stop := signal.NotifyContext(ctx, os.Interrupt, syscall.SIGTERM)
```
Comment (krtkvrm) [inline server/sys.go:11]: do you recommend also listening to `SIGINT` for graceful shutdowns? its basically when we do `ctrl + c `. 
Ref: https://dsa.cs.tsinghua.edu.cn/oj/static/unix_signal.html

File: server/grpc.go (hunk 4 above)
Comment (rohilsurana) [reply server/grpc.go:76]: We are using the implementation available in [grpc-go](https://github.com/grpc/grpc-go/tree/master/health)

File: go.mod (hunk 1 above)
Comment (rohilsurana) [reply go.mod:9]: This is a dependency from the added example. These are the once given as example in https://github.com/grpc-ecosystem/go-grpc-middleware package, so just used the same here.

File: server/mux.go (hunk 3 above)
Comment (rohilsurana) [reply server/mux.go:45]: The package provides an option for both use separate ports, or use on the same port. The reason allowing same port is that it enables users to be able to use both grpc and http without much changes in there infrastructure and deployments.

File: server/example/main.go (hunk 5 above)
Comment (rohilsurana) [reply server/example/main.go:103]: will refactor

File: server/http.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	server.httpMux = http.NewServeMux()
+
+	return server, nil
+}
+
+func (s *HTTPServer) Serve() error {
+	s.httpServer.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		s.httpMux.ServeHTTP(w, r)
+	})
+	return s.httpServer.ListenAndServe()
```
[the remark this answers is not in the document]
Comment (rohilsurana) [reply server/http.go:50]: Seems like a feature. Let's open a separate issue for the same?

File: server/example/main.go (hunk 5 above)
Comment (rohilsurana) [reply server/example/main.go]: Done.

File: server/http.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (krtkvrm) [reply server/http.go:50]: makes sense, for http its 1 line but would need to make some additions, that can make it a separate PR

File: server/gw.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+type GRPCGateway struct {
```
Comment (harikrishnakanchi) [inline server/gw.go:12]: Can we have a documentation/comment on GRPCGateway type. Helpful for consumers.

File: server/gw.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	// gwmux is the grpc-gateway proxy multiplexer
+	gwmux   *runtime.ServeMux
+	address string
+}
+
+func NewGateway(host string, port int) (*GRPCGateway, error) {
+	return &GRPCGateway{
+		gwmux: runtime.NewServeMux(
+			runtime.WithErrorHandler(runtime.DefaultHTTPErrorHandler),
```
Comment (harikrishnakanchi) [inline server/gw.go:21]: I believe this line is redundant. By default `runtime.NewServeMux` have default error handler as `runtime.DefaultErrorHandler`.

Comment (rohilsurana) [reply server/gw.go]: Done.

File: server/gw.go (hunk 8 above)
Comment (rohilsurana) [reply server/gw.go:13]: Added.

**Review by krtkvrm — APPROVED** [review]

**Review by harikrishnakanchi — APPROVED** [review]

**Review by ravisuhag — APPROVED** [review]
LGTM

**Pull request merged**